---
title: "01e_MROT_DataInput_Polyandry.rmd"
author: "Garett Slater"
date: "May 26 2022"
output:
  html_document: default
  pdf_document: default
---

```{r knitr setup, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
library(knitr)
knitr::opts_chunk$set(eval=TRUE, cache=FALSE, message=FALSE, warning=FALSE, 
                      comment = "")
```


```{r, include=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Code")
source("Polyandry_Evo_Functions_V2.R", local = knitr::knit_global())
```


# Data Input
```{r data}
########################################
##                                    ##       
##        Resource Files              ##
##                                    ## 
########################################
library(pacman)
p_load(biomatRt,tximport,DESeq2,dplyr,plyranges,ggplot2,plyr,ggpubr,reshape2,UpSetR,tispec,tidyverse,rstatix,readxl)

#Function Classes
func<-read.csv("~/OneDrive/Projects/Experimental Protocols/Bioinformatics/Proteomics/drone_function_final.csv")

########################################
##                                    ##       
## 1. Gene Datasets and Figure 1A     ##
##                                    ## 
########################################
########################################
##               BIMP                 ##
########################################
#Description Files
MROT_Sample = vector(mode = "list",length=3)
MROT_Sample[[1]]$Sample<-read.csv("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Resources/MROT_samp.csv")
MROT_Sample[[2]]$ortho<-read.csv("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Resources/MROT_ortho.csv")
MROT_Sample[[3]]$tx2gene<-read.csv("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Resources/MROT_tx2gene.csv")
```


# Gene Expression, Kallisto, and DESEq analysis
```{r echo=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Abundance")
#DESeq Analysis
#Run kallisto with Final (Tissue Type) as design and a count cutoff of 10
MROT_final = vector(mode = "list", length = 4)

#Gondal
Sample<-MROT_Sample[[1]]$Sample[which(MROT_Sample[[1]]$Sample$Tissue_Type=="Gonadal"),]
dds<-kallisto(Sample,MROT_Sample[[3]]$tx2gene,Type=~Final)
MROT_final[[1]]$Gonadal<-contrast("Final","MG","FG","MROT_")
```

# Gonadal Upset
```{r echo=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Graphics/RMD_Graphics")
MROT_final[[1]]$Gonadal$Bias[MROT_final[[1]]$Gonadal$MB == 1]<-"MG"
MROT_final[[1]]$Gonadal$Bias[MROT_final[[1]]$Gonadal$FB == 1]<-"FG"
MROT_final[[1]]$Gonadal$species<-"MROT"

pdf(file = "1-MROT-UpsetPlot-Gonadal.pdf", width = 12, height = 9);
par(cex = 1.6);
par(mar = c(0,4,2,0))
upset(MROT_final[[1]]$Gonadal[,2:3],order.by=c("freq"))
dev.off()
upset(MROT_final[[1]]$Gonadal[,2:3],order.by=c("freq"))

```

# Gonadal Ortho Upset
```{r echo=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Graphics/RMD_Graphics")
#Gondal-Ortho Only
MROT_final[[3]]$Gonadal_Ortho = merge(MROT_Sample[[2]]$ortho, MROT_final[[1]]$Gonadal,by="MROT_protein_id",all=TRUE)
MROT_final[[3]]$Gonadal_Ortho<-MROT_final[[3]]$Gonadal_Ortho %>%
  mutate(across(everything(), ~ ifelse(is.na(.), 0, .)))  %>%
  mutate(UB=ifelse(MB+FB==0,1,0))%>%
  filter(protein_id!=0) %>%
  select(protein_id,MB,FB,UB)

MROT_final[[3]]$Gonadal_Ortho$Bias[MROT_final[[3]]$Gonadal_Ortho$MB == 1]<-"MG"
MROT_final[[3]]$Gonadal_Ortho$Bias[MROT_final[[3]]$Gonadal_Ortho$FB == 1]<-"FG"
MROT_final[[3]]$Gonadal_Ortho$Bias[MROT_final[[3]]$Gonadal_Ortho$UB == 1]<-"UB"
MROT_final[[3]]$Gonadal_Ortho$species<-"MROT"

pdf(file = "1-MROT-UpsetPlot-GonadalOrtho.pdf", width = 12, height = 9);
par(cex = 1.6);
par(mar = c(0,4,2,0))
upset(MROT_final[[3]]$Gonadal_Ortho[,2:4],order.by=c("freq"))
dev.off()
upset(MROT_final[[3]]$Gonadal_Ortho[,2:4],order.by=c("freq"))

```

```{r echo=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Abundance")
#Somatic - Kallisto function and then contrast function
Sample<-MROT_Sample[[1]]$Sample[which(MROT_Sample[[1]]$Sample$Tissue_Type=="Somatic"),]
dds<-kallisto(Sample,MROT_Sample[[3]]$tx2gene,Type=~Final)
MROT_final[[2]]$Somatic<-contrast("Final","MS","FS","MROT_")
```

# Somatic Upset
```{r echo=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Graphics/RMD_Graphics")
MROT_final[[2]]$Somatic$Bias[MROT_final[[2]]$Somatic$MB == 2]<-"MS"
MROT_final[[2]]$Somatic$Bias[MROT_final[[2]]$Somatic$FB == 2]<-"FS"
MROT_final[[2]]$Somatic$species<-"MROT"

pdf(file = "1-MROT-UpsetPlot-Somatic.pdf", width = 12, height = 9);
par(cex = 1.6);
par(mar = c(0,4,2,0))
upset(MROT_final[[2]]$Somatic[,2:3],order.by=c("freq"))
dev.off()
upset(MROT_final[[2]]$Somatic[,2:3],order.by=c("freq"))

```

# Somatic Ortho Upset
```{r echo=FALSE}
setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Graphics/RMD_Graphics")
#Somatic-Ortho Only
MROT_final[[4]]$Somatic_Ortho = merge(MROT_Sample[[2]]$ortho, MROT_final[[2]]$Somatic,by="MROT_protein_id",all=TRUE)
MROT_final[[4]]$Somatic_Ortho<-MROT_final[[4]]$Somatic_Ortho %>%
  mutate(across(everything(), ~ ifelse(is.na(.), 0, .)))  %>%
  mutate(UB=ifelse(MB+FB==0,1,0)) %>%
  filter(protein_id!=0) %>%
  select(protein_id,MB,FB,UB)

MROT_final[[4]]$Somatic_Ortho$Bias[MROT_final[[4]]$Somatic_Ortho$MB == 1]<-"MS"
MROT_final[[4]]$Somatic_Ortho$Bias[MROT_final[[4]]$Somatic_Ortho$FB == 1]<-"FS"
MROT_final[[4]]$Somatic_Ortho$Bias[MROT_final[[4]]$Somatic_Ortho$UB == 1]<-"UB"
MROT_final[[4]]$Somatic_Ortho$species<-"MROT"

pdf(file = "1-MROT-UpsetPlot-SomaticOrtho.pdf", width = 12, height = 9);
par(cex = 1.6);
par(mar = c(0,4,2,0))
upset(MROT_final[[4]]$Somatic_Ortho[,2:4],order.by=c("freq"))
dev.off()
upset(MROT_final[[4]]$Somatic_Ortho[,2:4],order.by=c("freq"))

setwd("~/OneDrive/Manuscripts/Polyandry_GeneEvolution/Resources")
save(MROT_final,MROT_Sample,file = "MROT.RData")
```


